### ISSUE TITLE
[FSDP]  using CPUOffload creates 3-10x slowdown due to slow cpu optimizer step/update

### ISSUE THREAD

**Issue by lessw2020** [pytorch/pytorch#74588]
### 🐛 Describe the bug

Create simple distributed model
Wrapper model with FSDP. 
Using stateful optimizer ala Adam(W) run without CPUoffload and profile/time.
Then run with CPUOffload and see that performance has dramatically gone down, from 3x-10x vs without CPUOffloading.

Inspect profiling trace and see that slowdown is the optimizer update step (thanks @mrshenli!) and not D2H:

<img width="1047" alt="offloading_profile_trace" src="https://user-images.githubusercontent.com/46302957/159588512-d6464a4f-9e52-4fd0-b5da-d430c823ecc7.png">

FairSeq FSDP avoids this by wrappering CPUAdam from DeepSpeed.  
This uses a custom cpp AdamW build with AVX2 primitives and is thus 5-7x faster vs torch.optim.Adam. 
In addition, they have a fusion kernel to auto move the updated params back to GPU (claims +30% faster). 

The massive slowdown as-is basically means CPUOffloading is not usable with any adaptive optimizer. 

### Versions

Pytorch 1.11 stable and nightly. 
Reproed with and without FSDP custom wrapping. 
Slowdown varies from 3x-10x based on CPUs available (AWS g4dn, p3 etc). 

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by fduwjj** [comment]
Looks like this is FDSP specific and it's about the CPU offload feature. Do you know if it is possible that we can also do the similar stuff, @zhaojuanmao, @rohan-varma ?

**Comment by rohan-varma** [comment]
Thanks for raising this @lessw2020 and for pinging it @fduwjj!

Overall, yes we are aware of this issue and my understanding is that @mrshenli has been working with @lessw2020 on these benchmarks to evaluate CPU offload overhead with the optimizer.

We have a couple of proposals we are discussing internally which the core FSDP developers are aware of, and we can aim to put out an OSS RFC once the plans are more clear. @lessw2020 please ping if you'd like to be involved in these discussions!

At a high level, we're considering:
1. Implementing a performant CPU-based Adam optimizer
2. Utilizing https://github.com/pytorch/pytorch/issues/67570 to run optimizer as a callback of gradient communication. We've thought about doing this both on CPU and GPU as well, the latter would require careful management of optimizer states on the device.

**Comment by lessw2020** [comment]
Hi @rohan-varma  - I would love to be involved on the discussions for both approaches.  Thanks for the updates here and will ping you internally.

**Comment by ShinoharaHare** [comment]
Any update?
